### PR TITLE
Refactor Recipe and improve Recipe Step Graph Editor

### PR DESCRIPTION
**Pull request by seowalex**
# Things broken
* Downloading/publishing recipes seem to break the step graph
* Deleting recipes don't delete their associated images

# TODO
* Stop using exceptions as control flow

### CODE REVIEW

**Review by sc-arecrow — APPROVED** [review]
LGTM

File: ChopChop/ViewModels/Recipes/RecipeViewModel.swift [hunk 1]
```diff
@@ -1,121 +1,50 @@
...
-
-    private func bindPublished() {
-        recipe.$onlineId
-            .sink { [weak self] onlineId in
-                self?.isPublished = onlineId != nil
-            }
-            .store(in: &cancellables)
     }
 
+    // TODO: Properly throw errors
```
Comment (chrystalquek) [inline ChopChop/ViewModels/Recipes/RecipeViewModel.swift:29]: remove? some TODOs below as well

File: ChopChop/Views/Recipes/RecipeView.swift [hunk 2]
```diff
@@ -2,141 +2,158 @@ import SwiftUI
...
+                        Image(systemName: "square.and.pencil")
+                    }
+                }
+                Menu {
+                    Button(action: viewModel.publish) {
+                        Label(viewModel.isPublished ? "Publish changes" : "Publish", systemImage: "icloud.and.arrow.up")
+                    }
+
+                    // TODO: Make this work
+                    if viewModel.isPublished {
```
Comment (chrystalquek) [inline ChopChop/Views/Recipes/RecipeView.swift]: only unpublish from OnlineRecipe section

File: ChopChop/Views/Recipes/RecipeView.swift [hunk 3]
```diff
@@ -2,141 +2,158 @@ import SwiftUI
...
+            Divider()
+
+            if viewModel.isPublished {
+                Text("Published")
+                    .foregroundColor(.accentColor)
+            } else {
+                Text("Unpublished")
+                    .foregroundColor(.secondary)
             }
-            Text("Time taken: \(viewModel.totalTimeTaken)")
```
Comment (chrystalquek) [inline ChopChop/Views/Recipes/RecipeView.swift:86]: Do we still want to display this as an estimate?

**Review by chrystalquek — APPROVED** [review]

**Pull request merged**